packages/horizon/contracts/escrow/GraphEscrow.sol

```diff
@@ -0,0 +1,221 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.24;
+
+import { IGraphEscrow } from "../interfaces/IGraphEscrow.sol";
+import { IGraphPayments } from "../interfaces/IGraphPayments.sol";
+import { GraphDirectory } from "../GraphDirectory.sol";
+import { GraphEscrowStorageV1Storage } from "./GraphEscrowStorage.sol";
+import { TokenUtils } from "../utils/TokenUtils.sol";
+
+contract GraphEscrow is IGraphEscrow, GraphEscrowStorageV1Storage, GraphDirectory {
+    // -- Errors --
+
+    error GraphEscrowNotGraphPayments();
+    error GraphEscrowInputsLengthMismatch();
+    error GraphEscrowInsufficientThawAmount();
+    error GraphEscrowInsufficientAmount(uint256 available, uint256 required);
+    error GraphEscrowNotThawing();
+    error GraphEscrowStillThawing(uint256 currentTimestamp, uint256 thawEndTimestamp);
+    error GraphEscrowThawingPeriodTooLong(uint256 thawingPeriod, uint256 maxThawingPeriod);
+    error GraphEscrowCollectorNotAuthorized(address sender, address dataService);
+    error GraphEscrowCollectorInsufficientAmount(uint256 available, uint256 required);
+
+    // -- Events --
+
+    event AuthorizedCollector(address indexed sender, address indexed dataService);
+    event ThawCollector(address indexed sender, address indexed dataService);
+    event CancelThawCollector(address indexed sender, address indexed dataService);
+    event RevokeCollector(address indexed sender, address indexed dataService);
+    event Deposit(address indexed sender, address indexed receiver, uint256 amount);
+    event CancelThaw(address indexed sender, address indexed receiver);
+    event Thaw(
+        address indexed sender,
+        address indexed receiver,
+        uint256 amount,
+        uint256 totalAmountThawing,
+        uint256 thawEndTimestamp
+    );
+    event Withdraw(address indexed sender, address indexed receiver, uint256 amount);
+    event Collect(address indexed sender, address indexed receiver, uint256 amount);
+
+    // -- Constructor --
+
+    constructor(
+        address _controller,
+        uint256 _revokeCollectorThawingPeriod,
+        uint256 _withdrawEscrowThawingPeriod
+    ) GraphDirectory(_controller) {
+        if (_revokeCollectorThawingPeriod > MAX_THAWING_PERIOD) {
+            revert GraphEscrowThawingPeriodTooLong(_revokeCollectorThawingPeriod, MAX_THAWING_PERIOD);
+        }
+
+        if (_withdrawEscrowThawingPeriod > MAX_THAWING_PERIOD) {
+            revert GraphEscrowThawingPeriodTooLong(_withdrawEscrowThawingPeriod, MAX_THAWING_PERIOD);
+        }
+
+        revokeCollectorThawingPeriod = _revokeCollectorThawingPeriod;
+        withdrawEscrowThawingPeriod = _withdrawEscrowThawingPeriod;
+    }
+
+    // approve a data service to collect funds
+    function approveCollector(address dataService, uint256 amount) external {
```

Comment:
A payer could bypass the thawing by re-approving a collector for a 0 amount. 

I think we should not allow re approvals for amounts that are smaller than the current allowance. 

If you want to reduce the allowance it should go through the thawing period.